hnix-store-core/src/System/Nix/Nar.hs

```diff
@@ -0,0 +1,216 @@
+            str "entry"
+            parens $ do
+              str "name"
+              str (BSL.fromStrict $ E.encodeUtf8 name)
+              str "node"
+              parens (putFile fso)
+
+        parens m = str "(" >> m >> str ")"
+
+        str :: BSL.ByteString -> B.Put
+        str t = let len = BSL.length t
+            in int len <> pad t
+
+        int :: Integral a => a -> B.Put
+        int n = B.putInt64le $ fromIntegral n
+
+        pad :: BSL.ByteString -> B.Put
+        pad bs = do
+          B.putLazyByteString bs
+          B.putLazyByteString (BSL.replicate (padLen (BSL.length bs)) '\NUL')
+
+        strs :: [BSL.ByteString] -> B.Put
+        strs = mapM_ str
+
+
+------------------------------------------------------------------------------
+-- | Deserialize a Nar from lazy ByteString
+getNar :: B.Get Nar
+getNar = fmap Nar $ header >> parens getFile
+    where
+
+      header   = assertStr "nix-archive-1"
+
+
+      -- Fetch a FileSystemObject
+      getFile = getRegularFile <|> getDirectory <|> getSymLink
+
+      getRegularFile = do
+          assertStr "type"
+          assertStr "regular"
+          mExecutable <- optional $ Executable <$ (assertStr "executable"
+                                                   >> assertStr "")
+          assertStr "contents"
+          contents <- str
+          return $ Regular (fromMaybe NonExecutable mExecutable) contents
+
+      getDirectory = do
+          assertStr "type"
+          assertStr "directory"
+          fs <- many getEntry
+          return $ Directory (Map.fromList fs)
+
+      getSymLink = do
+          assertStr "type"
+          assertStr "symlink"
+          assertStr "target"
+          fmap SymLink str
+
+      getEntry = do
+          assertStr "entry"
+          parens $ do
+              assertStr "name"
+              mname <- E.decodeUtf8 . BSL.toStrict <$> str
+              assertStr "node"
+              file <- parens getFile
+              maybe (fail $ "Bad PathName: " ++ show mname)
+                    (return . (,file))
+                    (pathName mname)
+
+      -- Fetch a length-prefixed, null-padded string
+      str = do
+          n <- B.getInt64le
+          s <- B.getLazyByteString n
+          p <- B.getByteString . fromIntegral $ padLen n
+          return s
+
+      parens m = assertStr "(" *> m <* assertStr ")"
+
+      assertStr s = do
+          s' <- str
+          if s == s'
+              then return s
+              else fail "No"
+
+
+-- | Distance to the next multiple of 8
+padLen :: Int64 -> Int64
+padLen n = (8 - n) `mod` 8
+
+
+-- | Unpack a FileSystemObject into a non-nix-store directory (e.g. for testing)
+localUnpackNar :: FilePath -> Nar -> IO ()
+localUnpackNar basePath (Nar fso) = do
+
+  -- Create regular files and directories,
+  -- But save link creation until the end, by writing link commands out
+  -- to a queue, since otherwise we may try to create a link before
+  -- its target exists
```

Comment:
Wait, why does this matter? symlinks don't need to be valid, you can absolutely have a nar with `foo -> bar` but nothing at `bar`.

Reply:
The nar may contain invalid links, but we have to sequence creation of those links on the disk, since the trying to _create_ a link will fail if the target does not exist.

Or so I thought!! TIL that `ln -s path1 path2` does not check that `path1` exists! So I'll write `localUnpackNar` without the `WriterT` part.